webui/react/src/e2e/models/hew/DataGrid.ts

```diff
@@ -0,0 +1,330 @@
+  readonly defaultSelector: string = '[class^="DataGrid_base"]';
+  constructor(args: TableArgs<RowType, HeadRowType>) {
+    super(args);
+    this.#rowType = args.rowType;
+    this.rows = new args.rowType({
+      parent: this.#body,
+      parentTable: this,
+    });
+    this.headRow = new args.headRowType({
+      clickableParentLocator: this.pwLocator,
+      parent: this.#head,
+    });
+  }
+  readonly canvasTable: BaseComponent = new BaseComponent({
+    parent: this,
+    selector: 'canvas[data-testid="data-grid-canvas"] table',
+  });
+  readonly #otherCanvas: BaseComponent = new BaseComponent({
+    parent: this,
+    selector: 'canvas:not([data-testid])',
+  });
+  #columnheight: number | undefined;
+  readonly #rowType: RowClass<RowType, HeadRowType>;
+  readonly rows: RowType;
+  readonly headRow: HeadRowType;
+  readonly #body: BaseComponent = new BaseComponent({
+    parent: this.canvasTable,
+    selector: 'tbody',
+  });
+  readonly #head: BaseComponent = new BaseComponent({
+    parent: this.canvasTable,
+    selector: 'thead',
+  });
+
+  get columnHeight(): number {
+    if (this.#columnheight === undefined || Number.isNaN(this.#columnheight)) {
+      throw new Error('Please use setColumnHeight to set the column height');
+    }
+    return this.#columnheight;
+  }
+
+  /**
+   * Returns the height of the header column. Trying to click on a row will fail
+   * if this method hasn't yet been run.
+   */
+  async setColumnHeight(): Promise<number> {
+    const style = await this.#otherCanvas.pwLocator.getAttribute('style');
+    if (style === null) {
+      throw new Error("Couldn't find style attribute.");
+    }
+    const matches = style.match(/height: (\d+)px;/);
+    if (matches === null) {
+      throw new Error("Couldn't find height in style attribute.");
+    }
+    this.#columnheight = +matches[1];
+    return this.columnHeight;
+  }
+
+  /**
+   * Returns a row from an index. Start counting at 0.
+   */
+  getRowByIndex(n: number): RowType {
+    return new this.#rowType({
+      attachment: `[${this.rows.indexAttribute}="${n + 2}"]`,
+      parent: this.#body,
+      parentTable: this,
+    });
+  }
+
+  /**
+   * Returns a list of keys associated with attributes from rows from the entire table.
+   */
+  async allRows(): Promise<string[]> {
+    const { pwLocator, indexAttribute } = this.rows;
+    const rows = await pwLocator.all();
+    return Promise.all(
+      rows.map(async (row) => {
+        return (
+          (await row.getAttribute(indexAttribute)) ||
+          Promise.reject(new Error(`all rows should have the attribute ${indexAttribute}`))
+        );
+      }),
+    );
+  }
+
+  /**
+   * Returns a list of rows that match the condition provided
+   * @param {(row: RowType) => Promise<boolean>} condition - function which tests each row against a condition
+   */
+  async filterRows(condition: (row: RowType) => Promise<boolean>): Promise<RowType[]> {
+    return (
+      await Promise.all(
```

Comment:
This data grid stuff does not seem thread-safe. Not an issue now, but it probably will be later. If another test is creating/deleting experiments any index we are tracking goes out the window, and I think these multi-promise loops could wind up with contradictory data. If we have to use the index to interact with this, maybe we have a semaphore around experiments? I'm open to ideas hee too. I'm thinking it would be nice to have separate read/write locks. i.e.

- `Lock/UnlockWrite()`: prevents reading from experiment list and writing experiments. Test code writing to experiments would be expected to acquire this lock before making changes.
- `Lock/UnlockRead()`: prevents writing to experiments, but allows other tests to read. Tests reading experiment lists(like this one) would be expected to acquire this lock before reading.  

In other words,`LockWrite` would block `LockRead` and additional `LockWrite`s. `LockRead` would block `LockWrite`, but not additional `LockRead`s

Reply:
I think I would need to refresh the page before seeing any new experiments, right?

Reply:
no -- the experiment list requests the current page every 10 seconds

Reply:
useful! I think in that case the playwright locator would expire and throw an error. Maybe I could handle that error somehow. I left a todo since it's not in the scope of this ticket